proposals/0000-local-modules.rst

```diff
@@ -862,6 +868,26 @@ Each numbered item in this section can be considered separately.
    It is an error to omit a module's name and
    include the ``qualified`` keyword.
 
+   The original name of an entity declared within an anonymous local module
+   includes a special module name ``anonymous`` in place of the name of the
+   anonymous module. This special module name is guaranteed not to conflict
+   with any other module name, because it begins with a lower-case letter.
```

Comment:
The concept of an “original name” is only relevant to explain the semantics of this proposal, but the proposal does _not_ really dictate that as the implementation, nor does it show up in user-visible ways, right?

I am bit worried that it won’t hold water. This is one instance, but also code like
```
module Top
  module A where qualified module B where foo = True
  module qualified A where module B where foo = True
```
How do the two `foo`’s get different “original names”?

I am not sure if I can construct an example where there are two `foo` definitions with same original name, where I can still _use_ `foo` without getting an ambiguity error. But really, shouldn’t the semantics be able to explain even the steps done before such errors?

Ah, here is an example. I would expect this to work, but it seems that the two `foo`’s have clashing original names? 
```
module Top where
  module (bar) where foo = …; bar = 1 + foo
  module (baz) where foo = …; baz =1 + foo
```

Would it work to use the source position of the definition in the original name, instead of some path? Or simply an abstract per-file definition counter?

Reply:
Original names *do* matter to users, because if I am merging two QEnvs `{ x |-> orig1 }` and `{ x |-> orig2 }`, then I will get an unambiguous `x` if `orig1 == orig2`, and an ambiguous `x` otherwise. This is true today, where this can happen if you import the same definition that has been re-exported from multiple different imported modules.

Your first `foo` example is rejected: you cannot define two things with the same original name. This is also a logical consequence of the fact that modules can be extended. Indeed, we could refactor

```
module Local where
  x = 5
  y = True
```

to

```
module Local where x = 5
module Local where y = True
```

This refactoring would not be possible if `x` and `y` share some non-empty set of non-exported helper definitions, but thinking about the refactoring may be helpful.

Note that

```
module Local () where hidden = 5
module Local where x = hidden
```

does not work -- as I think should already be encoded in the proposal.

As for your later example: that idea flew across my mind, but it didn't find any comfortable perch, so it flew out the other end. We could either reject with "no two definitions can have the same original name" (actually, we'd need a better message) or say that anonymous modules actually are named `anonymous_N`, where `N` counts the number of anonymous modules from the top of the file, or something like that.

Blargh. Now a new thought just flew in.

```
module Top where

module M1 (a) where x = 5; a = x    -- x is not exported
module M1 (b) where x = 6; b = x
```

For anonymous modules, we can just rename. But for *named* modules, this is harder. My tendency is to reject with "don't do this", but I can actually see this restriction as being annoying. Maybe non-exported definitions from local modules get fancy original names? I don't really like that either.